Change the implementation design of matchLabelKeys in PodTopologySpread to be aligned with PodAffinity

mochizuki875 opened the pull request:
#### What type of PR is this?
/kind feature
/sig scheduling

#### What this PR does / why we need it:
As described in #129480, the design of `matchLabelKeys` is different between `podAffinity` and `topologySpreadConstraint`, which might confuse users.
This PR makes topologySpread's `matchLabelKeys` behave the same as podAffinity.

#### Which issue(s) this PR fixes:
Fixes #129480

#### Special notes for your reviewer:
I plan to include a description in the documentation stating [updating the labels is not supported](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades#the-update-to-labels-specified-at-matchlabelkeys-isnt-supported).

~https://github.com/kubernetes/website/pull/49602~
https://github.com/kubernetes/website/pull/49607

#### Does this PR introduce a user-facing change?

```release-note
Kubernetes api-server now merges selectors built from matchLabelKeys into the labelSelector of topologySpreadConstraints, 
aligning Pod Topology Spread with the approach used by Inter-Pod Affinity.

To avoid breaking existing pods that use matchLabelKeys, the current scheduler behavior will be preserved until it is removed in v1.34. 
Therefore, do not upgrade your scheduler directly from v1.32 to v1.34. 
Instead, upgrade step-by-step (from v1.32 to v1.33, then to v1.34), 
ensuring that any pods created at v1.32 with matchLabelKeys are either removed or already scheduled by the time you reach v1.34.

If you maintain controllers that previously relied on matchLabelKeys (for instance, to simulate scheduling), 
you likely no longer need to handle matchLabelKeys directly. Instead, you can just rely on the labelSelector field going forward.

Additionally, a new feature gate `MatchLabelKeysInPodTopologySpreadSelectorMerge`, which is enabled by default, has been 
added to control this behavior.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades
```

mochizuki875 committed:
Change the implementation design of matchLabelKeys in PodTopologySpread to be aligned with PodAffinity

mochizuki875 commented:
/cc @sanposhiho

sanposhiho commented:
/cc @macsko @dom4ha

sanposhiho commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7959,7 +7959,7 @@ func validateMatchLabelKeysAndMismatchLabelKeys(fldPath *field.Path, matchLabelK
 				// Here, we found the duplicate key in labelSelector and the key is specified in labelKeys.
 				// Meaning that the same key is specified in both labelSelector and matchLabelKeys/mismatchLabelKeys.
-				allErrs = append(allErrs, field.Invalid(fldPath.Index(i), key, "exists in both matchLabelKeys and labelSelector"))
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("matchLabelKeys").Index(i), key, "exists in both matchLabelKeys and labelSelector"))
```
Looks like a right change though, better submit another PR for this; it's essentially not related to the main course of this PR.

mochizuki875 replied:
Previously `validateMatchLabelKeysAndMismatchLabelKeys()` has been used for validating PodAffinity's `matchLabelKeys`/`mismatchLabelKeys`.
In this PR, it's also used for `TopologySpreadConstraints`.
Should I address only this in a separate PR initially?

sanposhiho replied:
Yes.

mochizuki875 replied:
OK, I've submitted #129900.

macsko commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -816,6 +817,26 @@ func mutatePodAffinity(pod *api.Pod) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.MatchLabelKeysInPodTopologySpread) || pod.Spec.TopologySpreadConstraints == nil {
+		return
+	}
+	if topologySpreadConstraint := pod.Spec.TopologySpreadConstraints; topologySpreadConstraint != nil {
+		for i := range topologySpreadConstraint {
+			applyMatchLabelKeys(&topologySpreadConstraint[i], pod.Labels)
+		}
+	}
+}
```
You already check `topologySpreadConstraint != nil` in the first condition, so you can simplify:
```suggestion
	topologySpreadConstraint := pod.Spec.TopologySpreadConstraints
	for i := range topologySpreadConstraint {
		applyMatchLabelKeys(&topologySpreadConstraint[i], pod.Labels)
	}
}
```

macsko commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -816,6 +817,26 @@ func mutatePodAffinity(pod *api.Pod) {
 	}
 }
 
+func applyMatchLabelKeys(constraint *api.TopologySpreadConstraint, label map[string]string) {
```
```suggestion
func applyMatchLabelKeys(constraint *api.TopologySpreadConstraint, labels map[string]string) {
```

macsko commented on pkg/registry/core/pod/strategy_test.go:
```diff
@@ -1789,6 +1789,214 @@ func Test_mutatePodAffinity(t *testing.T) {
+		name               string
+		pod                *api.Pod
+		wantPod            *api.Pod
+		featureGateEnabled bool
```
Maybe use more descriptive name, e.g. `matchLabelKeysEnabled`?

mochizuki875 commented:
I have submitted #129900 and will restart this PR once it is merged.

sanposhiho commented:
This PR isn't blocked by https://github.com/kubernetes/kubernetes/pull/129900, right?

mochizuki875 commented:
> This PR isn't blocked by https://github.com/kubernetes/kubernetes/pull/129900, right?

In this PR, `validateMatchLabelKeysAndMismatchLabelKeys()` is used to validate PodTopologySpread's matchLabelKeys, and the fixed result in #129900 is expected in the unit-test.

https://github.com/kubernetes/kubernetes/blob/39e23eda3d99afd02d1f6f9efd97dd482b91438a/pkg/apis/core/validation/validation_test.go#L22743

sanposhiho commented:
I mean, you can implement a unit test case with the previous behavior in this PR, and
- if this PR is merged before [#129900](https://github.com/kubernetes/kubernetes/pull/129900), [#129900](https://github.com/kubernetes/kubernetes/pull/129900) would have a failing unit test (that is added by this PR) and you can modify the test case in [#129900](https://github.com/kubernetes/kubernetes/pull/129900).
- on the contrary, if [#129900](https://github.com/kubernetes/kubernetes/pull/129900) is merged before this PR, your new test case here would start to fail in this PR and then you can modify here.

mochizuki875 commented:
@sanposhiho @macsko 
Thank you for your comment and I've addressed them, including [this comment](https://github.com/kubernetes/kubernetes/pull/129874#issuecomment-2623731141).
Please check again.

sanposhiho commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -22726,15 +22726,23 @@ func TestValidateTopologySpreadConstraints(t *testing.T) {
 			},
 		}},
-		wantFieldErrors: field.ErrorList{field.Invalid(fieldPathMatchLabelKeys.Index(0), "foo", "exists in both matchLabelKeys and labelSelector")},
+		// TODO: This expected result should be reverted after #129900 is merged.
+		// wantFieldErrors: field.ErrorList{field.Invalid(fieldPathMatchLabelKeys.Index(0), "foo", "exists in both matchLabelKeys and labelSelector")},
```
Generally speaking, comment out the implementation isn't a great idea. You can just say:

```suggestion
		// TODO: This expected message is not perfect, and will be fixed in #129900.
```

mochizuki875 replied:
> Generally speaking, comment out the implementation isn't a great idea.

Thank you for your advice.
BTW I have one question about your suggestion.
Is  the comment "will be fixed in #129900" appropriate?
This fix is not included in #129900 because #129900 has not be implemented PodTopologySpread's matchLabelKeys.

sanposhiho replied:
I don't quite get what you're trying to mean, but https://github.com/kubernetes/kubernetes/pull/129900 is gonna fix `validateMatchLabelKeysAndMismatchLabelKeys` which is used for Pod Topology Spread in this PR, right?
So, if this PR is merged before #129900, you ought to modify #129900 to delete this comment and fix the test case. If #129900 is merged before this PR, you can just remove this TODO and modify the test case.

mochizuki875 replied:
Oh, sorry, I was a bit confused...
Upon reconsidering, I understand it correctly now.

sanposhiho reviewed (commented):
Looking good overall

sanposhiho commented:
/release-note-edit
```release-note
Kubernetes api-server now merges selectors built from matchLabelKeys into the labelSelector of topologySpreadConstraints, 
aligning Pod Topology Spread with the approach used by Inter-Pod Affinity.

To avoid breaking existing pods that use matchLabelKeys, the current scheduler behavior will be preserved until it is removed in v1.34. 
Therefore, do not upgrade your scheduler directly from v1.32 to v1.34. 
Instead, upgrade step-by-step (from v1.32 to v1.33, then to v1.34), 
ensuring that any pods created at v1.32 with matchLabelKeys are either removed or already scheduled by the time you reach v1.34.

If you maintain controllers that previously relied on matchLabelKeys (for instance, to simulate scheduling), 
you likely no longer need to handle matchLabelKeys directly. Instead, you can just rely on the labelSelector field going forward.
```

mochizuki875 commented:
> Special notes for your reviewer:
I plan to include a description in the documentation stating [updating the labels is not supported](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades#the-update-to-labels-specified-at-matchlabelkeys-isnt-supported).

I've submitted PR.
https://github.com/kubernetes/website/pull/49602

macsko commented on pkg/registry/core/pod/strategy.go:
```diff
@@ -816,6 +817,25 @@ func mutatePodAffinity(pod *api.Pod) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.MatchLabelKeysInPodTopologySpread) || pod.Spec.TopologySpreadConstraints == nil {
+		return
+	}
+	topologySpreadConstraint := pod.Spec.TopologySpreadConstraints
```
nit: topologySpreadConstraint -> topologySpreadConstraints

mochizuki875 replied:
Thanks!

macsko commented:
/lgtm

sanposhiho reviewed (approved):
/lgtm

sanposhiho commented:
/assign @deads2k

mochizuki875 commented:
/retest

toVersus commented on pkg/registry/core/pod/strategy_test.go:
```diff
@@ -1789,6 +1789,214 @@ func Test_mutatePodAffinity(t *testing.T) {
+	}
+}
+
 func TestPodLifecycleSleepActionEnablement(t *testing.T) {
```
Maybe we could add a test to ensure the constraint explicitly stated in the following section of the KEP is properly enforced?
https://github.com/kubernetes/enhancements/blob/cc7b34e6c73c9cfa691e4bf13277f003df19a8f5/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades/README.md#L310

Something like this:

<details>
<summary>TestUpdateLabelOnPodWithTopologySpreadConstraintsEnabled</summary>


```go
func TestUpdateLabelOnPodWithTopologySpreadConstraintsEnabled(t *testing.T) {
	defaultTerminationGracePeriodSeconds := int64(30)

	pod := &api.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "foo",
			Namespace: "default",
			Labels: map[string]string{
				"country": "Japan",
				"city":    "Tokyo",
			},
		},
		Spec: api.PodSpec{
			RestartPolicy: api.RestartPolicyAlways,
			DNSPolicy:     api.DNSDefault,
			Containers: []api.Container{
				{
					Name:                     "container",
					Image:                    "image",
					ImagePullPolicy:          "IfNotPresent",
					TerminationMessagePolicy: "File",
				},
			},
			TopologySpreadConstraints: []api.TopologySpreadConstraint{
				{
					MaxSkew:           1,
					TopologyKey:       "kubernetes.io/hostname",
					WhenUnsatisfiable: api.DoNotSchedule,
					LabelSelector:     &metav1.LabelSelector{},
					MatchLabelKeys:    []string{"country", "city"},
				},
			},
			TerminationGracePeriodSeconds: &defaultTerminationGracePeriodSeconds,
		},
	}

	wantPod := &api.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "foo",
			Namespace: "default",
			Labels: map[string]string{
				"country": "Japan",
				"city":    "Tokyo",
				"ward":    "Shibuya",
			},
		},
		Spec: api.PodSpec{
			RestartPolicy: api.RestartPolicyAlways,
			DNSPolicy:     api.DNSDefault,
			Containers: []api.Container{
				{
					Name:                     "container",
					Image:                    "image",
					ImagePullPolicy:          "IfNotPresent",
					TerminationMessagePolicy: "File",
				},
			},
			TopologySpreadConstraints: []api.TopologySpreadConstraint{
				{
					MaxSkew:           1,
					TopologyKey:       "kubernetes.io/hostname",
					WhenUnsatisfiable: api.DoNotSchedule,
					LabelSelector: &metav1.LabelSelector{
						MatchExpressions: []metav1.LabelSelectorRequirement{
							{
								Key:      "country",
								Operator: metav1.LabelSelectorOpIn,
								Values:   []string{"Japan"},
							},
							{
								Key:      "city",
								Operator: metav1.LabelSelectorOpIn,
								Values:   []string{"Tokyo"},
							},
						},
					},
					MatchLabelKeys: []string{"country", "city"},
				},
			},
			TerminationGracePeriodSeconds: &defaultTerminationGracePeriodSeconds,
		},
	}

	featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.MatchLabelKeysInPodTopologySpread, true)
	Strategy.PrepareForCreate(genericapirequest.NewContext(), pod)
	if errs := Strategy.Validate(genericapirequest.NewContext(), pod); len(errs) != 0 {
		t.Errorf("Unexpected error: %v", errs.ToAggregate())
	}
	updatedPod := pod.DeepCopy()
	updatedPod.ObjectMeta.Labels["ward"] = "Shibuya"
	Strategy.PrepareForUpdate(genericapirequest.NewContext(), updatedPod, pod)
	if diff := cmp.Diff(wantPod.Spec.TopologySpreadConstraints, updatedPod.Spec.TopologySpreadConstraints); diff != "" {
		t.Errorf("unexpected modification to TopologySpreadConstraints. diff = %v", diff)
	}
}
```

</details>

mochizuki875 replied:
Thank you for your comment and code example.
I've incorporate your suggestion.

mochizuki875 commented:
I've added additional unit-test case for ensuring [KEP-3243 The update to labels specified at matchLabelKeys isn't supported](https://github.com/kubernetes/enhancements/blob/cc7b34e6c73c9cfa691e4bf13277f003df19a8f5/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades/README.md#the-update-to-labels-specified-at-matchlabelkeys-isnt-supported).

mochizuki875 commented:
Add origin information to fix the conflict with #130388.

macsko commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -22829,17 +22829,22 @@ func TestValidateTopologySpreadConstraints(t *testing.T) {
-			field.Invalid(fieldPathMatchLabelKeys.Index(0), nil, "").WithOrigin("overlappingKeys"),
-		},
+		// TODO: This expected message is not perfect, and will be fixed in #129900.
+		wantFieldErrors: field.ErrorList{field.Invalid(subFldPath0.Index(0), "foo", "exists in both matchLabelKeys and labelSelector").WithOrigin("overlappingKeys")},
```
It seems you shouldn't pass the detail string here, as error equality is now based on origin.
```suggestion
		wantFieldErrors: field.ErrorList{field.Invalid(subFldPath0.Index(0), "foo", "").WithOrigin("overlappingKeys")},
```

mochizuki875 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -8046,37 +8046,7 @@ func validateMatchLabelKeysAndMismatchLabelKeys(fldPath *field.Path, matchLabelK
-		allErrs = append(allErrs, unversionedvalidation.ValidateLabelName(key, fldPath.Index(i))...)
-		if labelSelectorKeys.Has(key) {
-			allErrs = append(allErrs, field.Invalid(fldPath.Index(i), key, "exists in both matchLabelKeys and labelSelector").WithOrigin("overlappingKeys"))
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("matchLabelKeys").Index(i), k, "exists in both matchLabelKeys and mismatchLabelKeys").WithOrigin("overlappingKeys"))
```
@macsko 
According to my understand, there are three cases of duplicating keys of `matchLabelKeys`:
- case1: `matchLabelKeys` and `labelSelector`
- case2: `matchLabelKeys` and `mismatchLabelKeys`
- case3: duplicating keys in `matchLabelKeys`
  - Currently, this case is not validated explicitly, but I think it should be done.(I've submitted #130554) 

While `overlappingKeys` is currently used for case1, what do you think about using a distinct origin as follows?
- case1: `duplicatedLabelKeys`
- case2: `duplicatedMismatchLabelKeys`
- case3: `duplicatedMatchLabelKeys`

sanposhiho commented:
/approve

sanposhiho commented:
@deads2k Can you take a look, or assign someone else if your bandwidth is limited?

deads2k commented:
It took me a while to chase down the various behavior and links to realize that this feature has been enabled by default with a different behavior for about two years.  At this point, we can reasonably expect that there are clients expecting the existing behavior.

I think I'm open to changing the behavior and allowing mutation based on this field, but it needs to be done under a different gate so that clusters negatively impacted by the change can disable it.  This new gate could be enabled by default and start as beta (same level as the current feature), but we cannot leave cluster stuck after an upgrade if this change to two year old behavior breaks them.

mochizuki875 commented:
@deads2k 

> but it needs to be done under a different gate so that clusters negatively impacted by the change can disable it. This new gate could be enabled by default and start as beta (same level as the current feature), but we cannot leave cluster stuck after an upgrade if this change to two year old behavior breaks them.

Thank you for the comments!
During the discussion of KEP-3243, we have considered the [risks and mitigations](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades#risks-and-mitigations) of this change, and concluded the risks are acceptably low.
In addition, we have also planned to provide the [safe upgrade path](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/3243-respect-pod-topology-spread-after-rolling-upgrades#risks-and-mitigations).
Do you think it's not sufficient and we need to prepare a new gate for controlling the mutation of key-value labels corresponding to `matchLabelKeys`?

deads2k commented:
> Do you think it's not sufficient and we need to prepare a new gate for controlling the mutation of key-value labels corresponding to `matchLabelKeys`?

Without a separate gate, a negatively impacted cluster is faced with an impossible choice of either
1. disable feature/MatchLabelKeysInPodTopologySpread, present since 1.27, which their cluster relies on
2. leave the new behavior in feature/MatchLabelKeysInPodTopologySpread enabled and suffer with negative behavior

To avoid cluster-admins being stuck in this position, we must introduce a featuregate for the pod label mutation and validation changes.  This new featuregate (which can start enabled by default and in beta) can be separately turned off so that a cluster that is negatively impacted by that mutation and validation can disable the new functionality, restore their cluster to the way it behaved in 1.32, and complain so we know the choice broke existing use-cases.

mochizuki875 commented:
@deads2k 
Thanks for the explanation.
But do we need to separate even the validation by featuregate?
In this PR, I've removed `validateMatchLabelKeysInTopologySpread()` and replaced it with `validateMatchLabelKeysAndMismatchLabelKeys()` because the validation logic is almost the same and doesn't become more strict. 
Furthermore, the current validation seems not to be controlled by the featuregate(`MatchLabelKeysInPodTopologySpread`).

Rather than maintaining two validations that are almost the same and controlling them with featuregates, I think it's better to consolidate them to one validation(`validateMatchLabelKeysAndMismatchLabelKeys()`) and continue without featuregate as before.

deads2k commented:
I'm glad you called it out.  I've looked through in more detail and I see these differences

1. old validation didn't check labelSelector when matchLabelKeys was empty,  new validation does check this
2. old validation doesn't allow labelSelector to be empty when matchLabelKeys was empty, new validation does allow this.
3. old validation checks all matchLabelKeys against keys in labelSelector.MatchLabel and labelSelector.MatcheExpressions.  New validation does something different, labelSelector.MatcheExpressions (only) keys are checked against matchLabelKeys AND labelSelector.MatchLabels at the same time.

That's a considerable list.  I'm actually worried about making this change at all.  Is it just beauty or is there a functional reason?

The previous validation wasn't gated because when the gate is off, the fields are empty.  With this gate, the fields will still be present and need to be validated differently.

mochizuki875 commented:
@sanposhiho  @macsko

Based on the feedback from @deads2k, we need to add a new FeatureGate.
Currently, [the milestone for this PR is set to v1.33](https://github.com/kubernetes/enhancements/issues/3243#issuecomment-2695028723), but since the enhancements freeze((02:00 UTC Friday 14th February 2025 / 19:00 PDT Thursday 13th February 2025)) has already passed, we need to change the milestone to v1.34.
If there is anything I misunderstand, I would appreciate it if you could point it out.

mochizuki875 commented:
@deads2k

Currently, `validateMatchLabelKeysAndMismatchLabelKeys()` is used for the validation of `PodAffinity`'s `matchLabelKeys`/`mismatchLabelKeys`, and the mutation of `labelSelector` corresponding to `matchLabelKeys`/`mismatchLabelKeys` is performed.
In this PR, the same mutation behavior of `matchLabelKeys` will be applied to `TopologySpreadConstraints`, so I think the requirement of the validation of `TopologySpreadConstraints`'s `matchLabelKeys` is the same as that of `PodAffinity`.

Especially:

> 3. old validation checks all matchLabelKeys against keys in labelSelector.MatchLabel and labelSelector.MatcheExpressions.  New validation does something different, labelSelector.MatcheExpressions (only) keys are checked against matchLabelKeys AND labelSelector.MatchLabels at the same time.

It comes from the difference that the mutation is performed or not.
If the mutation is enabled as the behavior of `PodAffinity`, the key-value labels corresponding to `matchLabelKeys` are added to `labelSelector` prior to the validation, and the old validation(`validateMatchLabelKeysInTopologySpread()`) cannot handle this case.
On the other hand, the new validation(`validateMatchLabelKeysAndMismatchLabelKeys()`) allows this case.

For these reasons, I changed it to use `validateMatchLabelKeysAndMismatchLabelKeys()` for the validation of `TopologySpreadConstraints`'s `matchLabelKeys` instead of `validateMatchLabelKeysInTopologySpread()`.

However, based on your list, I have reconsidered that it is better to separate the validations by FeatureGate as follows:
- Mutation is enabled(New FeatureGate is enabled): use `validateMatchLabelKeysAndMismatchLabelKeys()` to validate `matchLabelKeys`
- Mutation is disabled(New FeatureGate is disabled): use `validateMatchLabelKeysInTopologySpread()` to validate `matchLabelKeys`

deads2k commented:
> Based on the feedback from @deads2k, we need to add a new FeatureGate.
> Currently, [the milestone for this PR is set to v1.33](https://github.com/kubernetes/enhancements/issues/3243#issuecomment-2695028723), but since the enhancements freeze(02:00 UTC Friday 14th February 2025 / 19:00 PDT Thursday 13th February 2025) has already passed, we need to change the milestone to v1.34.
> If there is anything I misunderstand, I would appreciate it if you could point it out.

You can add a featuregate outside of KEP freeze.  The feature gate isn't changing the design, it is making the already agreed change less risky for all kubernetes clusters.  the KEP should be updated to reflect that it is gated, but since the design itself isn't changing you don't need to wait until 1.34 to merge the code.

mochizuki875 commented:
/release-note-edit

```
Kubernetes api-server now merges selectors built from matchLabelKeys into the labelSelector of topologySpreadConstraints, 
aligning Pod Topology Spread with the approach used by Inter-Pod Affinity.

To avoid breaking existing pods that use matchLabelKeys, the current scheduler behavior will be preserved until it is removed in v1.34. 
Therefore, do not upgrade your scheduler directly from v1.32 to v1.34. 
Instead, upgrade step-by-step (from v1.32 to v1.33, then to v1.34), 
ensuring that any pods created at v1.32 with matchLabelKeys are either removed or already scheduled by the time you reach v1.34.

If you maintain controllers that previously relied on matchLabelKeys (for instance, to simulate scheduling), 
you likely no longer need to handle matchLabelKeys directly. Instead, you can just rely on the labelSelector field going forward.

Additionally, a new feature gate `MatchLabelKeysInPodTopologySpreadMutation`, which is enabled by default, has been 
added to control this behavior.
```

mochizuki875 commented:
/release-note-edit

```release-note
Kubernetes api-server now merges selectors built from matchLabelKeys into the labelSelector of topologySpreadConstraints, 
aligning Pod Topology Spread with the approach used by Inter-Pod Affinity.

To avoid breaking existing pods that use matchLabelKeys, the current scheduler behavior will be preserved until it is removed in v1.34. 
Therefore, do not upgrade your scheduler directly from v1.32 to v1.34. 
Instead, upgrade step-by-step (from v1.32 to v1.33, then to v1.34), 
ensuring that any pods created at v1.32 with matchLabelKeys are either removed or already scheduled by the time you reach v1.34.

If you maintain controllers that previously relied on matchLabelKeys (for instance, to simulate scheduling), 
you likely no longer need to handle matchLabelKeys directly. Instead, you can just rely on the labelSelector field going forward.

Additionally, a new feature gate `MatchLabelKeysInPodTopologySpreadMutation`, which is enabled by default, has been 
added to control this behavior.
```

mochizuki875 commented:
I've submitted [the PR to update the KEP-3243](https://github.com/kubernetes/enhancements/pull/5205) and also reflected it in the PR.

mochizuki875 commented:
/retest

deads2k commented on pkg/apis/core/validation/validation.go:
```diff
@@ -7996,7 +7996,11 @@ func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstrai
 			allErrs = append(allErrs, err)
 		}
-		allErrs = append(allErrs, validateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
+		if utilfeature.DefaultFeatureGate.Enabled(features.MatchLabelKeysInPodTopologySpreadSelectorMerge) {
```
Instead of directly accessing the featuregates, use a new field on the `PodValidationOptions` (already passed to this method) and set in a spot similar to https://github.com/kubernetes/kubernetes/blob/07d66d9c268c7f8a52b7ba8c339bf21df1ec3ff1/pkg/api/pod/util.go#L405 .

The `allowLegacyTopologySpreadConstraints` also needs to consider whether the oldObject already violates the new validation in `validateMatchLabelKeysAndMismatchLabelKeys`, which we established is possible.  This is important so that we don't prevent unrelated updates to a pod.  In a case like

1. create with old validation, but fails new validation
2. attempt to delete (recall this is modeled as an update to handle finalizers), but fails new validation

we end up stuck unless we allow existing objects that don't pass new validation to be updated.

mochizuki875 replied:
@deads2k 
Thank you for your review comment.
I have one question.

Should we need to consider the reverse case?
As mentioned [here](https://github.com/kubernetes/kubernetes/pull/129874#issuecomment-2717005267), oldObject which is merged selector built from `matchLabelKeys` may violate the old validation(`validateMatchLabelKeysInTopologySpread()`).

> If the mutation is enabled as the behavior of PodAffinity, the key-value labels corresponding to matchLabelKeys are added to labelSelector prior to the validation, > and the old validation(validateMatchLabelKeysInTopologySpread()) cannot handle this case.
> On the other hand, the new validation(validateMatchLabelKeysAndMismatchLabelKeys()) allows this case.

So if users change the `MatchLabelKeysInPodTopologySpreadMutation` FeatureGate from true to false, could the similar case occur?

deads2k reviewed (changes requested):
Once we support existing and imperfect data I think this is ok.  Remember that both validation paths will exist ~forever when planning your client code.

mochizuki875 commented:
I've taken the comments into account.

mochizuki875 commented:
@deads2k 
PTAL?
Though I'm sorry for bothering you, the code freeze is approaching.
https://github.com/kubernetes/enhancements/issues/3243#issuecomment-2730269186

mochizuki875 commented:
/retest

deads2k commented on pkg/apis/core/validation/validation.go:
```diff
@@ -8011,7 +8019,17 @@ func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstrai
 			allErrs = append(allErrs, err)
 		}
-		allErrs = append(allErrs, validateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
+		if opts.AllowMatchLabelKeysInPodTopologySpread {
+			if opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge {
+				if !opts.AllowLegacyTopologySpreadConstraintsMatchLabelKeys {
+					allErrs = append(allErrs, ValidateMatchLabelKeysAndMismatchLabelKeys(subFldPath, constraint.MatchLabelKeys, nil, constraint.LabelSelector)...)
+				}
+			} else {
+				if !opts.AllowLegacyIncompatibleTopologySpreadConstraintsMatchLabelKeys {
+					allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
+				}
+			}
+		}
```
if `RequireValidateMatchLabelKeysAndMismatchLabelKeys` were the option and it was assigned like this in the strategy

```go
// we can require new validation if the old feature is enabled and the new feature merging is enabled and we don't have existing data that fails to validate.  If any of those are false, then we cannot require the new validation because it either isn't enabled/desired or the existing data doesn't work with it.
opts.RequireValidateMatchLabelKeysAndMismatchLabelKeys = utilfeature.DefaultFeatureGate.Enabled(features.MatchLabelKeysInPodTopologySpread) && utilfeature.DefaultFeatureGate.Enabled(features.MatchLabelKeysInPodTopologySpreadSelectorMerge) && !hasLegacyIncompatibleTopologySpreadConstrainMatchLabelKeys(oldSpec)
```

Then the code in here looks like

```go
if opts.RequireValidateMatchLabelKeysAndMismatchLabelKeys{
    allErrs = append(allErrs, ValidateMatchLabelKeysAndMismatchLabelKeys(subFldPath, constraint.MatchLabelKeys, nil, constraint.LabelSelector)...)
} else{
    allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
}
```

Which avoids the nesting. When the featuregates get locked to true, then the old validation will only be used for cases with existing bad data.

mochizuki875 replied:
@deads2k 
Thank you for your suggestion.

However, when each featuregate is enabled, the selector built from matchLabelKeys will be merged at the pod creation, and podSpec will no longer meet the old validation(`validateMatchLabelKeysInTopologySpread()`).
If my understanding is correct, this causes `opts.RequireValidateMatchLabelKeysAndMismatchLabelKeys` to be `false` because `hasLegacyIncompatibleTopologySpreadConstrainMatchLabelKeys(oldPodSpec)` returns `true`.
Is this appropriate?
I think the existing pod which meets the new validation(`ValidateMatchLabelKeysAndMismatchLabelKeys()`) can not be updated because it doesn't meet the old validation.

deads2k replied:
Good point.  I may have gotten that logic wrong. Backing up to what I'm trying to achieve: we need to validate with either old or new validation, so branches that don't end with validation are a problem and the code currently in the PR appears to have multiple ways to end up not validating at all.

I think it's logically this for create calls

```
if gate2Enabled{
  UseNewValidation
} else{
  UseOldValidation
}
```

For updates, I'm open to phrasings with a switch/case+default.   Are we able to do it with one gate and two checks of "which validation do you pass" (three fields total)?

```
switch{
case gate2Enabled && !oldPodPassesNewValidation:
  UseOldValidation
case gate2Enabled && oldPodPassesNewValidation:
  UseNewValidation
case !gate2Enabled && !oldPodPassesOldValidation:
  UseNewValidation
case !gate2Enabled && oldPodPassesOldValidation:
  UseOldValidation
default:
  // default specified so if we add another case, we'll always have something by default
  UseOldValidation
}
```

mochizuki875 replied:
Sorry for taking a little time to respond.
I've addressed.
Please check it again.

deads2k reviewed (commented):
it looks like this case avoids validation in some cases that coudl be enforced.  I've suggested an alternative that I think achieves the goal and validates with eitehr old or new.

deads2k commented on pkg/apis/core/validation/validation.go:
```diff
@@ -8055,7 +8063,20 @@ func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstrai
 		}
-		allErrs = append(allErrs, validateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
+		if opts.AllowMatchLabelKeysInPodTopologySpread {
+			switch {
+			case opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge && opts.OldPodViolatesMatchLabelKeysValidation:
+				allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
+			case opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge && !opts.OldPodViolatesMatchLabelKeysValidation:
+				allErrs = append(allErrs, ValidateMatchLabelKeysAndMismatchLabelKeys(subFldPath, constraint.MatchLabelKeys, nil, constraint.LabelSelector)...)
+			case !opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge && opts.OldPodViolatesLegacyMatchLabelKeysValidation:
+				allErrs = append(allErrs, ValidateMatchLabelKeysAndMismatchLabelKeys(subFldPath, constraint.MatchLabelKeys, nil, constraint.LabelSelector)...)
+			case !opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge && !opts.OldPodViolatesLegacyMatchLabelKeysValidation:
+				allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
+			default:
+				allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
+			}
```
For my future sanity it will help to add some comments.  Please review carefully to see if I've gotten it correct.

```suggestion

// legacyValidationFunction is  ValidateMatchLabelKeysInTopologySpread
// preferredValidationFunction is ValidateMatchLabelKeysAndMismatchLabelKeys
// OldPodViolatesMatchLabelKeysValidation==true means ValidateMatchLabelKeysAndMismatchLabelKeys failed, which means preferredValidation failed
// OldPodViolatesLegacyMatchLabelKeysValidation==true means ValidateMatchLabelKeysInTopologySpread failed


			switch {
			case opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge && opts.OldPodViolatesMatchLabelKeysValidation:
			// this case means that we want to use preferred validation, but the old pod doesn't pass new validation, so it must continue using legacyValidationFunction.  This is because we don't allow the fields to change.
				allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
			case opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge && !opts.OldPodViolatesMatchLabelKeysValidation:
			// this case means we want to use the preferred validation and the old pod worked with new validation, so we will continue requiring new the preferred validation to pass.
				allErrs = append(allErrs, ValidateMatchLabelKeysAndMismatchLabelKeys(subFldPath, constraint.MatchLabelKeys, nil, constraint.LabelSelector)...)
			case !opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge && opts.OldPodViolatesLegacyMatchLabelKeysValidation:
			// this is the case where we want legacy validation and the old pod does not pass legacy validation.  In this case we use the preferred (new) validation instead so that updates to other fields can happen. This allows us to enable the featuregate, then disable the featuregate and still be able to update the pod.
				allErrs = append(allErrs, ValidateMatchLabelKeysAndMismatchLabelKeys(subFldPath, constraint.MatchLabelKeys, nil, constraint.LabelSelector)...)
			case !opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge && !opts.OldPodViolatesLegacyMatchLabelKeysValidation:
			// this is the case where we want legacy validation and the old pod passes legacy validation.  In this case we use legacy validation
				allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
			default:
			// if we fall through, then we use legacy validation because that's what we did prior to the new gate.
				allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
			}
```

mochizuki875 replied:
Thank you for your suggestion!
I've addressed.

mochizuki875 committed:
fix1

mochizuki875 commented on pkg/apis/core/validation/validation.go:
```diff
@@ -8055,7 +8063,20 @@ func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstrai
+			default:
+				allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
+			}
+		}
```
I've added the else block and fixed one test case.

deads2k commented on pkg/features/kube_features.go:
```diff
@@ -1504,6 +1510,10 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
+	MatchLabelKeysInPodTopologySpreadSelectorMerge: {
+		{Version: version.MustParse("1.34"), Default: true, PreRelease: featuregate.Beta},
```
Noting for the record that this is adjustment to a feature already in beta, so going directly to beta and enabled by default is acceptable.  We are gating this so that if this adjustment causes negative consequences for a cluster, a cluster-admin can use the feature as it was before instead of being trapped.

deads2k commented:
/label tide/merge-method-squash
/lgtm
/approve

sanposhiho commented on pkg/apis/core/validation/validation.go:
```diff
@@ -8055,7 +8063,33 @@ func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstrai
 			allErrs = append(allErrs, err)
 		}
-		allErrs = append(allErrs, validateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
+		// legacyValidationFunction is ValidateMatchLabelKeysInTopologySpread
```
Wondering, in the first place though, do we need to have a validation on the update path? because all the related fields are immutable, right? So, instead of having those complexity, can we just skip validation on the update path?

mochizuki875 replied:
Considering [the comment](https://github.com/kubernetes/kubernetes/pull/129874#discussion_r2023412365) , I assumed all paths, including update path, must be validated.

> we need to validate with either old or new validation, so branches that don't end with validation are a problem and the code currently in the PR appears to have multiple ways to end up not validating at all.

@deads2k, do you have any comments on this?

sanposhiho replied:
I don't get that. What's the point of having the validation for the update path when you don't allow the field to be updated in the first place.

deads2k replied:
I didn't notice the immutability check. We do normally re-validate everything anyway, but if we enforce immutability I agree we could simplify.

mochizuki875 replied:
Thanks, so I'll review it and submit a PR.

sanposhiho commented on pkg/apis/core/validation/validation.go:
```diff
@@ -8055,7 +8063,33 @@ func validateTopologySpreadConstraints(constraints []core.TopologySpreadConstrai
+			case !opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge && !opts.OldPodViolatesLegacyMatchLabelKeysValidation:
+				// This case means we want to use the legacyValidationFunction and the old pod passes it, so we will continue requiring the legacyValidationFunction to pass.
+				allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
+			default:
+				// If we fall through, then we use the legacyValidationFunction because that's what we did prior to the featuregate(MatchLabelKeysInPodTopologySpreadSelectorMerge).
+				allErrs = append(allErrs, ValidateMatchLabelKeysInTopologySpread(subFldPath.Child("matchLabelKeys"), constraint.MatchLabelKeys, constraint.LabelSelector)...)
```
Can we reach here?

mochizuki875 replied:
Just like the else block right after this, it is to make it clear that validation is always be performed.
However, maybe it could have been combined with the case just above(`case !opts.AllowMatchLabelKeysInPodTopologySpreadSelectorMerge && !opts.OldPodViolatesLegacyMatchLabelKeysValidation`)...

sanposhiho replied:
The following `else` can be reached when `opts.AllowMatchLabelKeysInPodTopologySpread` is false, which is fine.
But, this `default` shouldn't be reached (iiuc), and we should remove it or we should, at least, log errors, not just returning something. (I do prefer removing it.)

mochizuki875 replied:
I've submit #132608.

sanposhiho reviewed (commented):
I intended to review the PR after https://github.com/kubernetes/enhancements/pull/5205 is merged though. Anyway..
